File: eidocell-ui/src/api/learning.ts
import { apiGet, apiPost, apiDelete } from './client'
import type {
  TrainModelRequest, TrainModelResult,
  RunInferenceRequest, InferenceResult,
  ApplyModelRequest, TrainedModelOut,
} from '@/types'

export const trainModel = (sid: string, data: TrainModelRequest) =>
  apiPost<TrainModelResult>(`/sessions/${sid}/learning/train`, data)

export const runInference = (sid: string, data: RunInferenceRequest) =>
  apiPost<InferenceResult>(`/sessions/${sid}/learning/infer`, data)

export const listSessionModels = (sid: string) =>
  apiGet<TrainedModelOut[]>(`/sessions/${sid}/learning/models`)

export const listAllModels = () =>
  apiGet<TrainedModelOut[]>('/models/')

export const getModel = (modelId: string) =>
  apiGet<TrainedModelOut>(`/models/${modelId}`)

export const deleteModel = (modelId: string) =>
  apiDelete(`/models/${modelId}`)

export const applyModel = (modelId: string, data: ApplyModelRequest) =>
  apiPost<InferenceResult>(`/models/${modelId}/apply`, data)

File: eidocell-backend/routers/learning.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session as DbSession

from db.session import get_db
from schemas.learning import (
    TrainModelRequest,
    TrainModelResult,
    RunInferenceRequest,
    InferenceResult,
    ApplyModelRequest,
    TrainedModelOut,
)
from services import learning_service

router = APIRouter(tags=["learning"])


# ── Session-scoped: train & infer ─────────────────────────────────────


@router.post(
    "/sessions/{session_id}/learning/train",
    response_model=TrainModelResult,
)
def train_model(
    session_id: str,
    data: TrainModelRequest,
    db: DbSession = Depends(get_db),
):
    return learning_service.train_model(
        db, session_id, data.name, data.classifier_type, data.feature_source, data.params
    )


@router.post(
    "/sessions/{session_id}/learning/infer",
    response_model=InferenceResult,
)
def run_inference(
    session_id: str,
    data: RunInferenceRequest,
    db: DbSession = Depends(get_db),
):
    return learning_service.run_inference(db, session_id, data.model_id)


@router.get(
    "/sessions/{session_id}/learning/models",
    response_model=list[TrainedModelOut],
)
def list_session_models(session_id: str, db: DbSession = Depends(get_db)):
    return learning_service.list_models(db, session_id=session_id)


# ── Global model registry ────────────────────────────────────────────


@router.get("/models/", response_model=list[TrainedModelOut])
def list_all_models(db: DbSession = Depends(get_db)):
    return learning_service.list_models(db)


@router.get("/models/{model_id}", response_model=TrainedModelOut)
def get_model(model_id: str, db: DbSession = Depends(get_db)):
    return learning_service.get_model(db, model_id)


@router.delete("/models/{model_id}", status_code=204)
def delete_model(model_id: str, db: DbSession = Depends(get_db)):
    learning_service.delete_model(db, model_id)


@router.post("/models/{model_id}/apply", response_model=InferenceResult)
def apply_model(
    model_id: str,
    data: ApplyModelRequest,
    db: DbSession = Depends(get_db),
):
    return learning_service.apply_model_cross_session(
        db, model_id, data.target_session_id
    )
